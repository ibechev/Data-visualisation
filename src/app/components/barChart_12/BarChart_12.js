import React from "react";
import PropTypes from "prop-types";

const BarChart_12 = ({ name, max, data }) => {
  const renderBars = () => {
    return (
      data &&
      Array.isArray(data) &&
      data.length === 12 &&
      data.map(({ value }, i) => {
        const barHeight = (value / max) * 100;

        return (
          <div key={i} className="bar">
            <span className="value" style={{ height: `${barHeight}%` }} />
          </div>
        );
      })
    );
  };

  return (
    <div className="bar-chart_12">
      <div className="chart-wrapper">{renderBars()}</div>
      <span className="bar-chart-label">{name}</span>
    </div>
  );
};

BarChart_12.propTypes = {
  name: PropTypes.string.isRequired,
  max(props, propName, component) {
    if (!props.hasOwnProperty(propName))
      return new Error(
        `Prop ${propName} at component ${component} is missing.`
      );

    if (typeof props[propName] !== "number")
      return new Error(
        `Prop ${propName} at component ${component} must be a number.`
      );

    if (props[propName] < 1)
      return new Error(
        `The value of prop ${propName} must be greater than 0. `
      );
  },
  data(props, propName) {
    return PropTypes.arrayOf(
      (propValue, key, component, location, propFullName) => {
        if (typeof propValue[key] !== "object")
          return new Error(
            `Prop ${propName} must be array of objects. Check ${propFullName}.`
          );

        if (propValue.length !== 12)
          return new Error(
            `Prop ${propName} at component ${component} must be array of length 12.`
          );

        if (
          !propValue[key].hasOwnProperty("month") ||
          !propValue[key].hasOwnProperty("value")
        )
          return new Error(
            `Each object in prop ${propName} at component ${component} must be with shape {month: "...", value: ...}. Check ${propFullName}..`
          );

        if (
          typeof propValue[key].month !== "string" ||
          !propValue[key].month.length
        )
          return new Error(
            `Key month of each object in prop ${propName} at component ${component} must be a string and must not be empty. Check ${propFullName}..`
          );

        if (
          typeof propValue[key].value !== "number" ||
          propValue[key].value < 0
        )
          return new Error(
            `Key value of each object in prop ${propName} at component ${component} must be a zero or a positive number. Check ${propFullName}..`
          );

        if (propValue[key]["value"] > props.max) {
          return new Error(
            `The value of key 'value' in each object of prop ${propName} should not exceed the value of prop max(${
              props.max
            }). Check ${propFullName}.`
          );
        }
      }
    ).isRequired.apply(this, arguments);
  }
};

export default BarChart_12;
