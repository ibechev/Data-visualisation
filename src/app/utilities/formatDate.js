export const formatDate = date => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ];
  const tempDate = date.split("-");
  return `${months[parseInt(tempDate[1]) - 1]} ${tempDate[0].substring(2)}`;
};
