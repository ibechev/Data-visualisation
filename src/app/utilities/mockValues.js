export const mockAlerts = [
  {
    name: "Adam Campton",
    event: "end of year review",
    eventDate: "2019-01-18",
    importance: "very"
  },
  {
    name: "Emma O'Sullivan",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Brandon Walsh",
    event: "end of year review",
    eventDate: "2019-02-06",
    importance: "middle"
  },
  {
    name: "David O'Brien",
    event: "end of year review",
    eventDate: "2019-02-23",
    importance: "very"
  },
  {
    name: "Sophie O'Connor",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Sean Connor",
    event: "end of year review",
    eventDate: "2019-01-05",
    importance: "very"
  },
  {
    name: "Gery Brown",
    event: "end of year review",
    eventDate: "2019-01-12",
    importance: "middle"
  },
  {
    name: "Stephen Greenday",
    event: "end of year review",
    eventDate: "2019-02-11",
    importance: "very"
  },
  {
    name: "Lisa Aniston",
    event: "end of year review",
    eventDate: "2018-12-27",
    importance: "low"
  },
  {
    name: "Gordon Niels",
    event: "end of year review",
    eventDate: "2018-12-28",
    importance: "middle"
  }
];

export const mockCustomerInfo = {
  max: 270,
  data: [
    { month: "Dec-2017", value: 99 },
    { month: "Jan-2018", value: 80 },
    { month: "Feb-2018", value: 71 },
    { month: "Mar-2018", value: 96 },
    { month: "Apr-2018", value: 80 },
    { month: "May-2018", value: 115 },
    { month: "Jun-2018", value: 146 },
    { month: "Jul-2018", value: 121 },
    { month: "Aug-2018", value: 160 },
    { month: "Sep-2018", value: 168 },
    { month: "Oct-2018", value: 185 },
    { month: "Nov-2018", value: 203 }
  ]
};
