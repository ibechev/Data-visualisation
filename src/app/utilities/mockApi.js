import { mockAlerts, mockCustomerInfo } from "../utilities/mockValues";

const randomTimeRange = (min, max) => {
  return (Math.floor(Math.random() * (max - min + 1)) + min) * 1000;
};
const simulateLatency = ms => new Promise(resolve => setTimeout(resolve, ms));

export const mockFetchAlerts = async () => {
  await simulateLatency(randomTimeRange(0, 3));
  return mockAlerts.map(alert => ({ ...alert }));
};

export const mockFetchClientsInfo = async () => {
  const { max, data } = mockCustomerInfo;
  await simulateLatency(randomTimeRange(0, 3));
  return {
    max,
    data: data.map(each => ({ ...each }))
  };
};
