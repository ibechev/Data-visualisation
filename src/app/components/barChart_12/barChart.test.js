import React from "react";
import { mount } from "enzyme";
import BarChart_12 from "./BarChart_12";
import renderer from "react-test-renderer";
import { propTypeErrors } from "../../../../tests/config/testUtils";
import { mockCustomerInfo } from "../../utilities/mockValues";

describe("Expect <BarChart />", () => {
  const defaultProps = {
    name: "Test name",
    max: mockCustomerInfo.max,
    data: [...mockCustomerInfo.data]
  };

  const wrapper = mount(<BarChart_12 {...defaultProps} />);

  it("to render correctly", () => {
    const snapshot = renderer
      .create(<BarChart_12 {...defaultProps} />)
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it("to not throw error with correct props", () => {
    propTypeErrors(<BarChart_12 {...defaultProps} />, defaultProps);
  });

  it("to render 12 bars", () => {
    expect(wrapper.find("div.bar")).toHaveLength(12);
  });

  it("to set the correct height of each bar", () => {
    let expectedHeight = "";
    wrapper.find("span.value").forEach((span, i) => {
      expectedHeight = (defaultProps.data[i].value / defaultProps.max) * 100;
      expect(span.prop("style")).toHaveProperty("height", `${expectedHeight}%`);
    });
  });

  it("to display prop 'name' indside <span class='bar-chart-label' />", () => {
    expect(wrapper.find("span.bar-chart-label").text()).toEqual(
      defaultProps.name
    );
  });
});
