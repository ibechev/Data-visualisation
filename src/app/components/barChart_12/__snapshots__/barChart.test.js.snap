// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <BarChart /> to render correctly 1`] = `
<div
  className="bar-chart_12"
>
  <div
    className="chart-wrapper"
  >
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "36.666666666666664%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "29.629629629629626%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "26.296296296296294%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "35.55555555555556%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "29.629629629629626%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "42.592592592592595%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "54.074074074074076%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "44.81481481481481%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "59.25925925925925%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "62.22222222222222%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "68.51851851851852%",
          }
        }
      />
    </div>
    <div
      className="bar"
    >
      <span
        className="value"
        style={
          Object {
            "height": "75.18518518518519%",
          }
        }
      />
    </div>
  </div>
  <span
    className="bar-chart-label"
  >
    Test name
  </span>
</div>
`;
